wptherml/optdriver.py

```diff
@@ -0,0 +1,1204 @@
+
+
+        def eval(self, num_eval = 20):
+            #for testing the inference accuracy of the model after training 
+
+            self.model.eval()
+            with torch.no_grad():
+                # Test out inference with 5 samples from the original test set
+                for i in range(num_eval):
+                    X_sample = self.X_test_raw[i: i+1]
+
+                    if self.standardscale or self.minmaxscale:
+                        y_pred = self.scaler.inverse_transform(y_pred)
+
+                    X_sample = torch.tensor(X_sample, dtype=torch.float32)
+                    y_pred = self.model(X_sample)
+                    if self.standardscale or self.minmaxscale:
+                        y_pred = self.scaler.inverse_transform(y_pred)
+
+                    print(f"{self.X_test_raw[i]} -> {y_pred[0]} (expected {self.y_test[i]})")
+
+
+
+        def get_weights(self):
+            #get weights from model as list
+            model_params = copy.deepcopy(self.best_weights)
+            return_list =  [np.array(model_params['lin.bias']), np.array(model_params['lin.weight']), np.array(model_params['V'])]
+            return return_list
+
+        def return_weights(self):
+            #get weights from model
+            params = self.get_weights()
+            V = np.array( np.array(params[2]).tolist())
+            W = np.array(list(list(np.array(params[1]))[0]), dtype = np.float64)
+            bias = np.array(params[0], dtype = np.float64)[0]
+
+            return V, W, bias
+
+
+        #following 2 methods
+        #https://github.com/tsudalab/fmbqm/blob/master/fmbqm/factorization_machine.py
+
+        def triu_mask(self, input_size, F=np.ndarray):
+            #Generate a square matrix with its upper trianguler elements being 1 and others 0.
+
+            mask = np.expand_dims(np.arange(input_size), axis=0)
+            return (np.transpose(mask) < mask) * 1.0
+
+        def VtoQ(self, V, F=np.ndarray):
+            """Calculate interaction strength by inner product of feature vectors.
+            """
+            print(V.shape)
+            #input_size = V.shape[1]
+            input_size = V.shape[0]
+
+
+            #Q = np.dot(np.transpose(V), V) # (d,d)
+            Q = np.dot(V , np.transpose(V)) # (d,d)
+            #Q =   V @ V.T
+            triu = self.triu_mask(input_size)
+            #print(triu)
+            #return Q * np.triu(np.ones((input_size,input_size)), 1)
+            return Q * triu
+            
+
+
+        def get_qubo_hamiltonian_minima(self):
+            """convert factorization machine to qubo hamiltonian and solve the model, getting one result"""
+
+            params = copy.deepcopy(self.get_weights())
+
+            V = np.array(params[2], dtype = np.float64)
+
+            W = np.array(np.array(params[1])[0], dtype = np.float64)
+
+            bias = np.array(params[0], dtype = np.float64)[0]
+
+
+            Q = V @ V.T
+
+
+            Q = self.VtoQ(V)
```

Comment:
Q = V @ V.T on line 1073 is what I would do, do you know why multiplying by the upper triangle matrix (triu) is necessary in the VtoQ method?  

Also right now you are computing Q twice, once in line 1073 and again in 1076